src/common/framework/gpu/device_pool.ts

```diff
@@ -64,66 +69,183 @@ export class DevicePool {
       // Any error that isn't explicitly TestFailedButDeviceReusable forces a new device to be
       // created for the next test.
       if (!(ex instanceof TestFailedButDeviceReusable)) {
-        this.holder = undefined;
+        if (holder === this.defaultHolder) {
+          this.defaultHolder = 'uninitialized';
+        } else {
+          this.nonDefaultHolders.deleteByDevice(holder.device);
+        }
+        // TODO: device.destroy()
       }
       throw ex;
     } finally {
-      // TODO: device.destroy()
-
       // Mark the holder as free. (This only has an effect if the pool still has the holder.)
       // This could be done at the top but is done here to guard against async-races during release.
-      holder.acquired = false;
+      holder.state = 'free';
     }
   }
+}
+
+interface DescriptorToDevice {
+  key: GPUDeviceDescriptor;
+  value: DeviceHolder;
+}
+
+/**
+ * Map from GPUDeviceDescriptor to DeviceHolder.
+ */
+class DescriptorToDeviceMap {
+  private unsupported: GPUDeviceDescriptor[] = [];
+  // TODO: Do something like stringifyPublicParamsUniquely if searching this array gets too slow.
+  private items: Set<DescriptorToDevice> = new Set();
+
+  /** Deletes an item from the map by GPUDevice value. */
+  deleteByDevice(device: GPUDevice): void {
+    for (const item of this.items) {
+      if (item.value.device === device) {
+        this.items.delete(item);
+        return;
+      }
+    }
+  }
+
+  /**
+   * Gets a DeviceHolder from the map if it exists; otherwise, calls create() to create one,
+   * inserts it, and returns it.
+   *
+   * Throws SkipTestCase if devices with this descriptor are unsupported.
+   */
+  async getOrInsert(
+    key: GPUDeviceDescriptor,
+    create: () => Promise<DeviceHolder>
+  ): Promise<DeviceHolder> {
+    // Never retry unsupported configurations.
+    for (const desc of this.unsupported) {
```

Comment:
Eeek O(n) is a bit sad, but IDK if there's a better way to do it in JS.

Reply:
It can be done like I did #289 recently. It doesn't have to be perfect anyway, it just is intended to reduce the number of redundant failed device requests. (It already isn't perfect because it doesn't know the difference between default values and unspecified values.)